pkg/addonmanager/controllers/registration/controller.go

```diff
@@ -155,23 +155,28 @@ func (c *addonRegistrationController) sync(ctx context.Context, syncCtx factory.
 		_, err = addonPatcher.PatchStatus(ctx, managedClusterAddonCopy, managedClusterAddonCopy.Status, managedClusterAddon.Status)
 		return err
 	}
-	configs := registrationOption.CSRConfigurations(managedCluster)
-
-	managedClusterAddonCopy.Status.Registrations = configs
-
-	managedClusterAddonCopy.Status.Namespace = registrationOption.Namespace
-	if len(managedClusterAddonCopy.Spec.InstallNamespace) > 0 {
-		managedClusterAddonCopy.Status.Namespace = managedClusterAddonCopy.Spec.InstallNamespace
-	}
 
+	var agentInstallNamespace string
 	if registrationOption.AgentInstallNamespace != nil {
-		ns, err := registrationOption.AgentInstallNamespace(managedClusterAddonCopy)
+		agentInstallNamespace, err = registrationOption.AgentInstallNamespace(managedClusterAddonCopy)
 		if err != nil {
+			if _, patchErr := addonPatcher.PatchStatus(ctx, managedClusterAddonCopy, managedClusterAddonCopy.Status, managedClusterAddon.Status); patchErr != nil {
+				return patchErr
+			}
 			return err
 		}
-		if len(ns) > 0 {
-			managedClusterAddonCopy.Status.Namespace = ns
-		}
+	}
+
+	configs := registrationOption.CSRConfigurations(managedCluster)
+	managedClusterAddonCopy.Status.Registrations = configs
+
+	// Determine the namespace, prioritizing in the order: agentInstallNamespace > InstallNamespace > Namespace
+	if len(agentInstallNamespace) > 0 {
+		managedClusterAddonCopy.Status.Namespace = agentInstallNamespace
+	} else if len(managedClusterAddonCopy.Spec.InstallNamespace) > 0 {
+		managedClusterAddonCopy.Status.Namespace = managedClusterAddonCopy.Spec.InstallNamespace
+	} else {
```

Comment:
I think we can default namespace to registrationOption.Namespace instead of else.

Reply:
fixed